### ISSUE TITLE
Installation issue: OPENMPI%clang@15.0.0

### ISSUE THREAD

**Issue by cjobrie** [LLNL/spack#40232]
### Steps to reproduce the issue

```console
$ spack install openmpi@4.1.5%apple-clang@15.0.0
...
```
This issue is likely related to #40221 and possibly #40223.

There are known issues with the new apple compiler that changes the linker implementation. I've tried to stay at v14.X, but my system continually updates xcode,  many times now, whenever I look away. 

The error is in linking with gfortran:

```console
==> Installing openmpi-4.1.5-v2ctg6tc62xaxw7okoe4nux6hae3fr4j
==> No binary for openmpi-4.1.5-v2ctg6tc62xaxw7okoe4nux6hae3fr4j found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/a6/a640986bc257389dd379886fdae6264c8cfa56bc98b71ce3ae3dfbd8ce61dbe3.tar.bz2
==> No patches needed for openmpi
==> openmpi: Executing phase: 'autoreconf'
==> openmpi: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/var/folders/d1/k4j62bm13657bj8x48w_l09r002j5p/T/cjobrie/spack-stage/spack-stage-openmpi-4.1.5-v2ctg6tc62xaxw7okoe4nux6hae3fr4j/spack-src/configure' '--prefix=/Users/cjobrie/alegra_project/toolset2/opt/nevada/darwin-ventura-skylake/apple-clang-15.0.0/openmpi-4.1.5-v2ctg6tc62xaxw7okoe4nux6hae3fr4j' '--enable-shared' '--disable-silent-rules' '--disable-builtin-atomics' '--enable-static' '--enable-mpi1-compatibility' '--without-verbs' '--without-ucx' '--without-ofi' '--without-fca' '--without-knem' '--without-mxm' '--without-psm2' '--without-cma' '--without-hcoll' '--without-xpmem' '--without-psm' '--without-cray-xpmem' '--without-sge' '--without-lsf' '--without-slurm' '--without-tm' '--without-loadleveler' '--without-alps' '--disable-memchecker' '--with-libevent=/Users/cjobrie/alegra_project/toolset2/opt/nevada/darwin-ventura-skylake/apple-clang-15.0.0/libevent-2.1.12-bbd55n562svmqpebqzi7spgigmprls3x' '--with-pmix=/Users/cjobrie/alegra_project/toolset2/opt/nevada/darwin-ventura-skylake/apple-clang-15.0.0/pmix-4.2.3-yequf546ydxtqig3j43sq5jnkwkefa5z' '--with-zlib=/Users/cjobrie/alegra_project/toolset2/opt/nevada/darwin-ventura-skylake/apple-clang-15.0.0/zlib-1.2.13-zowcqelilmocemidq5g2m6uiccu4tvfu' '--with-hwloc=/Users/cjobrie/alegra_project/toolset2/opt/nevada/darwin-ventura-skylake/apple-clang-15.0.0/hwloc-2.9.1-ut63grkaevk7dbbh7xxmkdin6wnq3bmt' '--disable-java' '--disable-mpi-java' '--with-gpfs=no' '--without-cuda' '--enable-wrapper-rpath' '--disable-wrapper-runpath' '--disable-mpi-cxx' '--disable-cxx-exceptions'

2 errors found in build log:
     371    checking the name lister (/usr/bin/nm -B) interface... BSD nm
     372    checking for fgrep... /usr/bin/grep -F
     373    checking if .proc/endp is needed... no
     374    checking directive for setting text section... .text
     375    checking directive for exporting symbols... .globl
     376    checking for objdump... objdump
  >> 377    checking if .note.GNU-stack is needed... /Library/Developer/CommandLineTools/usr/bin/objdump: error: 'c
            onftest.o': Invalid/Unsupported object file format
     378    no
     379    checking suffix for labels... :
     380    checking prefix for global symbol labels... _
     381    checking prefix for lsym labels... L
     382    checking prefix for function in .type...
     383    checking if .size is needed... no

     ...

     403    * execute.  Note that this is likely not a problem with Open MPI,
     404    * but a problem with the local compiler installation.  More
     405    * information (including exactly what command was given to the
     406    * compiler and what error resulted when the command was executed) is
     407    * available in the config.log file in the Open MPI build directory.
     408    **********************************************************************
  >> 409    configure: error: Could not run a simple Fortran program.  Aborting.
```

I've followed the guidance from Apple's [Xcode 15 release notes](https://developer.apple.com/documentation/xcode-release-notes/xcode-15-release-notes) by adding the linker flags to my spack compiler YAML file as:

```yaml
  compilers:
  - compiler:
      spec: apple-clang@15.0.0
      paths:
        cc: /usr/bin/clang
        cxx: /usr/bin/clang++
        f77: /usr/local/bin/gfortran-13
        fc: /usr/local/bin/gfortran-13
      flags:
        ldflags: -Wl,-ld_classic
      operating_system: ventura
      target: x86_64
      modules: []
      environment: {}
      extra_rpaths: []
```

However, this doesn't help OpenMPI@4.1.5 to compile. 

@hppritcha  @naughtont3

### Error message

<details><summary>Error message</summary>

<pre>
==> Installing openmpi-4.1.5-v2ctg6tc62xaxw7okoe4nux6hae3fr4j
==> No binary for openmpi-4.1.5-v2ctg6tc62xaxw7okoe4nux6hae3fr4j found: installing from source
==> Warning: upstream not found: /opt/spack/opt/spack/.spack-db/index.json
==> Fetching https://mirror.spack.io/_source-cache/archive/a6/a640986bc257389dd379886fdae6264c8cfa56bc98b71ce3ae3dfbd8ce61dbe3.tar.bz2
==> No patches needed for openmpi
==> openmpi: Executing phase: 'autoreconf'
==> openmpi: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/var/folders/d1/k4j62bm13657bj8x48w_l09r002j5p/T/cjobrie/spack-stage/spack-stage-openmpi-4.1.5-v2ctg6tc62xaxw7okoe4nux6hae3fr4j/spack-src/configure' '--prefix=/Users/cjobrie/alegra_project/toolset2/opt/nevada/darwin-ventura-skylake/apple-clang-15.0.0/openmpi-4.1.5-v2ctg6tc62xaxw7okoe4nux6hae3fr4j' '--enable-shared' '--disable-silent-rules' '--disable-builtin-atomics' '--enable-static' '--enable-mpi1-compatibility' '--without-verbs' '--without-ucx' '--without-ofi' '--without-fca' '--without-knem' '--without-mxm' '--without-psm2' '--without-cma' '--without-hcoll' '--without-xpmem' '--without-psm' '--without-cray-xpmem' '--without-sge' '--without-lsf' '--without-slurm' '--without-tm' '--without-loadleveler' '--without-alps' '--disable-memchecker' '--with-libevent=/Users/cjobrie/alegra_project/toolset2/opt/nevada/darwin-ventura-skylake/apple-clang-15.0.0/libevent-2.1.12-bbd55n562svmqpebqzi7spgigmprls3x' '--with-pmix=/Users/cjobrie/alegra_project/toolset2/opt/nevada/darwin-ventura-skylake/apple-clang-15.0.0/pmix-4.2.3-yequf546ydxtqig3j43sq5jnkwkefa5z' '--with-zlib=/Users/cjobrie/alegra_project/toolset2/opt/nevada/darwin-ventura-skylake/apple-clang-15.0.0/zlib-1.2.13-zowcqelilmocemidq5g2m6uiccu4tvfu' '--with-hwloc=/Users/cjobrie/alegra_project/toolset2/opt/nevada/darwin-ventura-skylake/apple-clang-15.0.0/hwloc-2.9.1-ut63grkaevk7dbbh7xxmkdin6wnq3bmt' '--disable-java' '--disable-mpi-java' '--with-gpfs=no' '--without-cuda' '--enable-wrapper-rpath' '--disable-wrapper-runpath' '--disable-mpi-cxx' '--disable-cxx-exceptions'

2 errors found in build log:
     371    checking the name lister (/usr/bin/nm -B) interface... BSD nm
     372    checking for fgrep... /usr/bin/grep -F
     373    checking if .proc/endp is needed... no
     374    checking directive for setting text section... .text
     375    checking directive for exporting symbols... .globl
     376    checking for objdump... objdump
  >> 377    checking if .note.GNU-stack is needed... /Library/Developer/CommandLineTools/usr/bin/objdump: error: 'c
            onftest.o': Invalid/Unsupported object file format
     378    no
     379    checking suffix for labels... :
     380    checking prefix for global symbol labels... _
     381    checking prefix for lsym labels... L
     382    checking prefix for function in .type...
     383    checking if .size is needed... no

     ...

     403    * execute.  Note that this is likely not a problem with Open MPI,
     404    * but a problem with the local compiler installation.  More
     405    * information (including exactly what command was given to the
     406    * compiler and what error resulted when the command was executed) is
     407    * available in the config.log file in the Open MPI build directory.
     408    **********************************************************************
  >> 409    configure: error: Could not run a simple Fortran program.  Aborting.

</pre></details>

Also, in the `config.log` i see:
<pre><config.log>
configure:44177: checking if Fortran compiler works
configure:44206: /Users/cjobrie/alegra_project/toolset2/third_party/spack/lib/spack/env/clang/gfortran -o conftest    -L/Users/cjobrie/alegra_project/toolset2/opt/nevada/darwin-ventura-skylake/apple-clang-15.0.0/zlib-1.2.13-zowcqelilmocemidq5g2m6uiccu4tvfu/lib  conftest.f  -lz >&5
ld: library not found for -ld64
collect2: error: ld returned 1 exit status
configure:44206: $? = 1
configure: program exited with status 1
configure: failed program was:
|       program main
| 
|       end
configure:44222: result: no
configure:44236: error: Could not run a simple Fortran program.  Aborting.
</pre></details>
### Information on your system

* **Spack:** 0.20.1 (2df338660c37158945c8c94c9fe7e42ed3d68e58)
* **Python:** 3.11.4
* **Platform:** darwin-ventura-skylake
* **Concretizer:** clingo
* apple-clang 15.0.0 (c/c++)
* gfortran 13 from homebrew

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/12744082/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/12744083/spack-build-out.txt)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by cpraveen** [comment]
I face the same issue. There are too many other packages that break. And Apple wont allow me to install clang14 on sonoma.

**Comment by cpraveen** [comment]
Adding

```
      flags:
        ldflags: -ld64
```

I succeeded to compile openmpi@4.1.5.

**Comment by cjobrie** [comment]
@cpraveen I've also tried adding those flag in the compiler section, but doesn't seem to help me out.

**Comment by cpraveen** [comment]
It worked on

intel MBP, sonoma, clang 15
M1 MBP, sonoma, clang 15

Since you are on ventura you can use xcode 14, maybe you need to disable automatic upgrading.

**Comment by cjobrie** [comment]
I've tried adding `-ld64` and (subsequently) `-Wl,-ld_classic` but that doesn't help either.

**Comment by cjobrie** [comment]
I don't think it's a openmpi issue specifically. I believe it has to do with SPACK's gfortran warpper not taking `-Wl,-ld_classic` as an argument, but rather processing it as a library! 

```shell 
configure:44177: checking if Fortran compiler works
configure:44206: /Users/cjobrie/alegra_project/toolset2/third_party/spack/lib/spack/env/clang/gfortran -o conftest    -L/Users/cjobrie/alegra_project/toolset2/opt/nevada/darwin-ventura-skylake/apple-clang-15.0.0/zlib-1.2.13-6m43uxc3u7olhtql36bhkg367ye3n4xw/lib  conftest.f  -lz >&5
ld: library not found for -ld_classic
collect2: error: ld returned 1 exit status
configure:44206: $? = 1
configure: program exited with status 1
configure: failed program was:
|       program main
|
|       end
configure:44222: result: no
configure:44236: error: Could not run a simple Fortran program.  Aborting.
```

**Comment by cpraveen** [comment]
Openmpi news says xcode 15 configure issue has been fixed in 4.1.6

https://raw.githubusercontent.com/open-mpi/ompi/v4.1.x/NEWS

**Comment by cjobrie** [comment]
@cpraveen you are correct that the latest openmpi@main fixes the configure error with clang@15.0.0. However, there is no package in spack for openmpi@4.1.6, so one would have to build with @main for now. 

Thanks for your help!
